File: src/go/downloader_sdk/downloader_sdk.go
// Package downloader_sdk is the Go equivalent of src/js/downloader-sdk.ts:
// a music player SDK that downloads files in chunks (strictly one chunk at a
// time per URL), starts playback as soon as the first chunk lands, supports
// seeking (reusing got chunks, keeping only the inflight request for the next
// needed chunk) and preloading (which always defers to an active playback).
//
// JS/Go mapping: AbortSignal -> context.Context, Promise -> blocking call
// returning error, injected playback promise -> PlaybackFunc. Unlike the JS
// version (which fires playback promises concurrently in index order),
// playback here is dispatched by a per-session goroutine strictly in order,
// which matches how an audio pipeline consumes chunks.
package downloader_sdk

import (
	"context"
	"errors"
	"sync"
	"time"
)

// GetFunc downloads one chunk of url. It must honor ctx cancellation.
type GetFunc func(ctx context.Context, url string) error

// PlaybackFunc plays one downloaded chunk.
type PlaybackFunc func(idx int) error

type Meta struct {
	Gets  []GetFunc
	Start int
}

// ErrPlaybackInterrupted settles a session that was replaced by a newer
// StartPlayback or SeekTo call.
var ErrPlaybackInterrupted = errors.New("playback interrupted")

// Fixed short retry delay (no exponential backoff): a stalled playback chunk
// is latency-critical, so fail fast and let the caller degrade instead of
// waiting out long backoffs.
const (
	defaultMaxRetries = 3
	defaultRetryDelay = 100 * time.Millisecond
)

type getter struct {
	done   chan struct{}
	err    error // written before done is closed
	cancel context.CancelFunc
	got    bool
}

type playSession struct {
	id      int
	url     string
	resCh   chan error
	playCh  chan int
	need    int
	played  int
	settled bool
}

type SDK struct {
	mu       sync.Mutex
	cond     *sync.Cond // signaled on download fires and session settles
	playback PlaybackFunc
	// Chunk cache keyed by url, then by absolute chunk index, so preloading
	// one url never corrupts the download state of another.
	cache   map[string]map[int]*getter
	session int
	current *playSession

	maxRetries int
	retryDelay time.Duration
}

func New() *SDK {
	s := &SDK{
		playback:   func(int) error { return nil },
		cache:      map[string]map[int]*getter{},
		maxRetries: defaultMaxRetries,
		retryDelay: defaultRetryDelay,
	}
	s.cond = sync.NewCond(&s.mu)
	return s
}

// SetPlayback injects the host player's chunk playback function.
func (s *SDK) SetPlayback(fn PlaybackFunc) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.playback = fn
}

func (s *SDK) chunksLocked(url string) map[int]*getter {
	m := s.cache[url]
	if m == nil {
		m = map[int]*getter{}
		s.cache[url] = m
	}
	return m
}

func (s *SDK) getWithRetry(ctx context.Context, get GetFunc, url string) error {
	for attempt := 0; ; attempt++ {
		err := get(ctx, url)
		if err == nil {
			return nil
		}
		// Aborted downloads (e.g. by a seek) must not be retried.
		if ctx.Err() != nil || attempt >= s.maxRetries {
			return err
		}
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(s.retryDelay):
		}
	}
}

func (s *SDK) fireDownloadLocked(url string, meta Meta, idx int) *getter {
	ctx, cancel := context.WithCancel(context.Background())
	g := &getter{done: make(chan struct{}), cancel: cancel}
	s.chunksLocked(url)[idx] = g
	get := meta.Gets[idx]
	go func() {
		g.err = s.getWithRetry(ctx, get, url)
		close(g.done)
	}()
	s.cond.Broadcast()
	return g
}

func (s *SDK) markGot(g *getter) {
	s.mu.Lock()
	if !g.got {
		g.got = true
		g.cancel = nil
	}
	s.mu.Unlock()
}

func (s *SDK) settleLocked(ps *playSession, err error) {
	if ps.settled {
		return
	}
	ps.settled = true
	ps.resCh <- err // buffered, single settle: never blocks
	if s.current == ps {
		s.current = nil
	}
	s.cond.Broadcast()
}

// StartPlayback downloads url in chunks starting at meta.Start and plays each
// chunk as soon as it lands. It blocks until every chunk from Start on has
// played, a download or playback fails, or a newer session replaces this one
// (ErrPlaybackInterrupted).
func (s *SDK) StartPlayback(url string, meta Meta) error {
	s.mu.Lock()
	// Starting a new playback replaces (and interrupts) any previous session.
	if s.current != nil {
		s.settleLocked(s.current, ErrPlaybackInterrupted)
	}
	s.session++
	start := max(meta.Start, 0)
	if start >= len(meta.Gets) {
		s.mu.Unlock()
		return nil
	}
	ps := &playSession{
		id:    s.session,
		url:   url,
		resCh: make(chan error, 1),
		need:  len(meta.Gets) - start,
	}
	ps.playCh = make(chan int, ps.need)
	s.current = ps
	s.mu.Unlock()

	go s.dispatchPlayback(ps)
	go s.runSession(url, meta, start, ps)
	return <-ps.resCh
}

// dispatchPlayback plays chunks strictly in the order they were downloaded,
// without ever blocking the download loop.
func (s *SDK) dispatchPlayback(ps *playSession) {
	for idx := range ps.playCh {
		s.mu.Lock()
		fn := s.playback
		stale := ps.settled || ps.id != s.session
		s.mu.Unlock()
		if stale {
			return
		}
		err := fn(idx)
		s.mu.Lock()
		if ps.settled || ps.id != s.session {
			s.mu.Unlock()
			return
		}
		if err != nil {
			s.settleLocked(ps, err)
			s.mu.Unlock()
			return
		}
		ps.played++
		if ps.played == ps.need {
			s.settleLocked(ps, nil)
		}
		s.mu.Unlock()
	}
}

func (s *SDK) runSession(url string, meta Meta, start int, ps *playSession) {
	defer close(ps.playCh)
	for idx := start; idx < len(meta.Gets); idx++ {
		s.mu.Lock()
		if ps.settled || ps.id != s.session {
			s.mu.Unlock()
			return
		}
		chunks := s.chunksLocked(url)
		g := chunks[idx]
		if g == nil {
			// The session's next chunk has priority: cancel any other
			// inflight download (e.g. a preload further ahead) so only one
			// chunk downloads at a time.
			for i, other := range chunks {
				if !other.got && i != idx {
					if other.cancel != nil {
						other.cancel()
					}
					delete(chunks, i)
				}
			}
			g = s.fireDownloadLocked(url, meta, idx)
		}
		got := g.got
		s.mu.Unlock()
		if !got {
			<-g.done
			s.mu.Lock()
			if ps.settled || ps.id != s.session {
				// A seek replaced this session while we were downloading; a
				// stale session must not touch shared cache entries.
				s.mu.Unlock()
				return
			}
			if g.err != nil {
				if chunks[idx] == g {
					delete(chunks, idx)
				}
				s.settleLocked(ps, g.err)
				s.mu.Unlock()
				return
			}
			g.got = true
			g.cancel = nil
			s.mu.Unlock()
		}
		ps.playCh <- idx
	}
}

// SeekTo restarts playback of url at chunk idx. Got chunks replay instantly
// without re-downloading. Among inflight requests, only the one for the first
// chunk at or after idx that still needs downloading is kept (and reused);
// every other inflight request is aborted so that chunk can start
// immediately. Like StartPlayback, it blocks until the new session settles.
func (s *SDK) SeekTo(url string, meta Meta, idx int) error {
	if idx < 0 || idx >= len(meta.Gets) {
		return nil
	}
	s.mu.Lock()
	chunks := s.chunksLocked(url)

	// Got chunks may be sparse (earlier seeks/preloads), so scan for the
	// first chunk at or after idx that still needs downloading.
	nextNeeded := idx
	for nextNeeded < len(meta.Gets) {
		g := chunks[nextNeeded]
		if g == nil || !g.got {
			break
		}
		nextNeeded++
	}

	for i, g := range chunks {
		if !g.got && i != nextNeeded {
			if g.cancel != nil {
				g.cancel()
			}
			// Drop the aborted entry so a later session re-downloads it.
			delete(chunks, i)
		}
	}
	s.mu.Unlock()

	m := meta
	m.Start = idx
	return s.StartPlayback(url, m)
}

func findInflightLocked(chunks map[int]*getter) *getter {
	for _, g := range chunks {
		if !g.got {
			return g
		}
	}
	return nil
}

// Preload downloads chunks of url sequentially starting at idx. It always
// defers to any inflight download (e.g. an active playback session) so only
// one chunk downloads at a time, and backs off (returning the error) when a
// download it depends on fails or is aborted by a seek.
func (s *SDK) Preload(url string, meta Meta, idx int) error {
	i := max(idx, 0)
	for {
		s.mu.Lock()
		if i >= len(meta.Gets) {
			s.mu.Unlock()
			return nil
		}
		chunks := s.chunksLocked(url)
		if g := chunks[i]; g != nil {
			if g.got {
				i++
				s.mu.Unlock()
				continue
			}
			// Someone else (playback or an earlier preload) is downloading it.
			s.mu.Unlock()
			<-g.done
			if g.err != nil {
				return g.err
			}
			s.markGot(g)
			i++
			continue
		}
		if inf := findInflightLocked(chunks); inf != nil {
			// Defer to the other inflight download, then re-check chunk i
			// since the state may have changed while waiting.
			s.mu.Unlock()
			<-inf.done
			if inf.err != nil {
				return inf.err
			}
			s.markGot(inf)
			continue
		}
		if cur := s.current; cur != nil && !cur.settled && cur.url == url {
			// An active playback session owns this url's download slot and
			// will fire its next chunk momentarily; wait instead of racing
			// it (it would cancel our request anyway).
			s.cond.Wait()
			s.mu.Unlock()
			continue
		}
		g := s.fireDownloadLocked(url, meta, i)
		s.mu.Unlock()
		<-g.done
		s.mu.Lock()
		if g.err != nil {
			if chunks[i] == g {
				delete(chunks, i)
			}
			s.mu.Unlock()
			return g.err
		}
		g.got = true
		g.cancel = nil
		s.mu.Unlock()
		i++
	}
}

File: chucks/js/downloader-sdk.ts
/**
 * # Music Player SDK Design and Implementation
 *
 * ## Background
 * We are developing a music player SDK that requires the following core features:
 *
 * 1. **Playback Functionality**: The `startPlayback` method accepts a URL and other
 *    required metadata to start chunked downloading and playback of the music file.
 *    To ensure a good user experience, playback should start as quickly as possible,
 *    and each downloaded chunk should be parsed and played as soon as it is available.
 *
 * 2. **Seek Functionality**: The `seekTo` method allows jumping to a specified position
 *    in the currently playing music, ensuring the correct chunk is downloaded and
 *    played.
 *
 * 3. **Preloading Functionality**: The `preload` method supports downloading chunks of
 *    a specified URL for faster startup during playback.
 *
 * ## Special Requirements
 * 1. **Chunked Downloading Strategy**: All music files must be downloaded in chunks.
 *    Only one chunk can be downloaded at a time, and playback can only start after the
 *    first chunk is fully downloaded.
 *
 * ## Requirements
 * Based on the above needs, please complete the following tasks:
 *
 * 1. **API Design**
 *    - Design the SDK interface and implement core logics on download.
 */

interface Meta {
  gets: ((url: string, signal: AbortSignal) => Promise<void>)[];
  start: number;
}

interface Getter {
  p: Promise<void> | null;
  abort: ((reason?: unknown) => void) | null;
  got: boolean;
}

// Chunk cache keyed by url, then by absolute chunk index, so preloading one
// url never corrupts the download state of another.
const cache = new Map<string, Map<number, Getter>>();

const chunksOf = (url: string): Map<number, Getter> => {
  let chunks = cache.get(url);
  if (!chunks) {
    chunks = new Map();
    cache.set(url, chunks);
  }
  return chunks;
};

let __resolve: ((value: void | PromiseLike<void>) => void) | null = null;
let __reject: ((reason?: unknown) => void) | null = null;
// Incremented whenever a new playback session starts; stale download loops
// compare against it and bail out instead of racing the new session.
let __session = 0;

const clearCbs = () => {
  __resolve = null;
  __reject = null;
};

type PlaybackFn = (idx: number) => Promise<void>;

// Injected by the host player; defaults to a no-op so the SDK is inert
// until wired up.
let playback: PlaybackFn = () => Promise.resolve();

function setPlayback(fn: PlaybackFn) {
  playback = fn;
}

// Fixed short retry delay (no exponential backoff): a stalled playback chunk
// is latency-critical, so fail fast and let the caller degrade instead of
// waiting out long backoffs.
const MAX_RETRIES = 3;
const RETRY_DELAY_MS = 100;

const sleep = (ms: number, signal: AbortSignal) =>
  new Promise<void>((res, rej) => {
    const onAbort = () => {
      clearTimeout(timer);
      rej(signal.reason ?? new Error("aborted"));
    };
    const timer = setTimeout(() => {
      signal.removeEventListener("abort", onAbort);
      res();
    }, ms);
    signal.addEventListener("abort", onAbort, { once: true });
  });

async function getWithRetry(
  get: Meta["gets"][number],
  url: string,
  signal: AbortSignal,
): Promise<void> {
  for (let attempt = 0; ; attempt++) {
    try {
      return await get(url, signal);
    } catch (err) {
      // Aborted downloads (e.g. by a seek) must not be retried.
      if (signal.aborted || attempt >= MAX_RETRIES) {
        throw err;
      }
      await sleep(RETRY_DELAY_MS, signal);
    }
  }
}

function fireDownload(url: string, metadata: Meta, idx: number): Getter {
  const chunks = chunksOf(url);
  const controller = new AbortController();
  const getter: Getter = {
    p: getWithRetry(metadata.gets[idx], url, controller.signal),
    abort: controller.abort.bind(controller),
    got: false,
  };
  chunks.set(idx, getter);
  return getter;
}

function markGot(url: string, idx: number) {
  const getter = chunksOf(url).get(idx);
  if (getter) {
    getter.p = null;
    getter.abort = null;
    getter.got = true;
  }
}

function interruptSession() {
  __reject?.(new Error("playback interrupted"));
  clearCbs();
}

function startPlayback(url: string, metadata: Meta): Promise<void> {
  // Starting a new playback replaces (and interrupts) any previous session.
  interruptSession();
  return new Promise<void>((res, rej) => {
    __resolve = res;
    __reject = rej;
    const session = ++__session;
    const { gets } = metadata;
    const start = Math.max(0, metadata.start);
    if (start >= gets.length) {
      res();
      clearCbs();
      return;
    }

    const chunks = chunksOf(url);
    const need = gets.length - start;
    let played = 0;
    let settled = false;

    const settle = (err?: unknown) => {
      settled = true;
      if (err === undefined) {
        res();
      } else {
        rej(err);
      }
      if (session === __session) {
        clearCbs();
      }
    };

    (async () => {
      // Chunks are downloaded strictly one at a time, but playback of each
      // chunk is fired as soon as it lands and never blocks the next download.
      for (let idx = start; idx < gets.length; idx++) {
        if (settled || session !== __session) {
          return;
        }
        let getter = chunks.get(idx);
        if (!getter || (!getter.got && !getter.p)) {
          getter = fireDownload(url, metadata, idx);
        }
        if (!getter.got) {
          try {
            await getter.p;
          } catch (err) {
            // Only the live session may clean up and settle; a stale session
            // (replaced by a seek) must not touch shared cache entries.
            if (session === __session && !settled) {
              chunks.delete(idx);
              settle(err);
            }
            return;
          }
          if (session !== __session || settled) {
            // A seek replaced this session while we were downloading.
            return;
          }
          markGot(url, idx);
        }
        playback(idx)
          .then(() => {
            played++;
            if (played === need && session === __session && !settled) {
              settle();
            }
          })
          .catch((err) => {
            if (session === __session && !settled) {
              settle(err);
            }
          });
      }
    })();
  });
}

function seekTo(
  url: string,
  metadata: Meta,
  idx: number,
): Promise<void> | void {
  const { gets } = metadata;
  if (idx < 0 || idx >= gets.length) {
    return;
  }
  const chunks = chunksOf(url);

  // Got chunks may be sparse (earlier seeks/preloads), so scan for the first
  // chunk at or after idx that still needs downloading.
  let nextNeeded = idx;
  while (nextNeeded < gets.length && chunks.get(nextNeeded)?.got) {
    nextNeeded++;
  }

  // Keep an inflight request only if it is exactly the next chunk we need
  // (the new session reuses its promise); any other inflight download would
  // block it (one chunk at a time) and is aborted so nextNeeded can start
  // immediately.
  for (const [i, g] of chunks) {
    if (!g.got && i !== nextNeeded) {
      g.abort?.();
      // Drop the aborted entry so a later session re-downloads this chunk.
      chunks.delete(i);
    }
  }

  return startPlayback(url, { ...metadata, start: idx });
}

function findInflight(chunks: Map<number, Getter>): Getter | undefined {
  for (const g of chunks.values()) {
    if (g.p) {
      return g;
    }
  }
  return undefined;
}

async function preload(url: string, metadata: Meta, idx: number) {
  const { gets } = metadata;
  const chunks = chunksOf(url);
  let i = Math.max(0, idx);
  while (i < gets.length) {
    const existing = chunks.get(i);
    if (existing?.got) {
      i++;
      continue;
    }
    if (existing?.p) {
      // Someone else (playback or an earlier preload) is downloading it.
      try {
        await existing.p;
      } catch {
        // Aborted by a seek or failed after retries: stop preloading rather
        // than racing the new session.
        return;
      }
      i++;
      continue;
    }
    // Only one chunk may download at a time per url: defer to any other
    // inflight download (e.g. an active playback session), then re-check
    // this index since the state may have changed while waiting.
    const inflight = findInflight(chunks);
    if (inflight?.p) {
      try {
        await inflight.p;
      } catch {
        return;
      }
      continue;
    }
    const getter = fireDownload(url, metadata, i);
    try {
      await getter.p;
    } catch {
      chunks.delete(i);
      return;
    }
    markGot(url, i);
    i++;
  }
}

export { startPlayback, seekTo, preload, setPlayback };
export type { Meta };
